common/build/eslint-plugin-fluid/src/rules/no-unchecked-record-access.js

```diff
@@ -302,86 +301,150 @@ function isUndefinableIndexSignatureType(parserServices, node) {
-function propertyHasBeenChecked(node) {
+/**
+ * Traverses up the AST from a property access node to check if the property has been properly guarded against being undefined.
+ * Looks for safety checks like optional chaining, null checks, 'in' operator usage, etc.
+ *
+ * @param {MemberExpression} node - The AST node representing the property access to check
+ * @param {RuleContext} context - ESLint rule context containing scope and AST information
+ * @returns {boolean} True if the property access has been properly checked for undefined, false otherwise
+ */
+function propertyHasBeenChecked(node, context) {
+	const baseObj = getBaseObject(node);
+	const currentKeyNode = node.property;
 	let current = node;
+
 	while (current) {
-		if (isDefined(current)) {
+		if (
+			current.optional || // Check for optional chaining (?.)
+			current.type === "ChainExpression" || // Check for nullish coalescing operator (??)
+			current.type === "TSNonNullExpression" // Check for non-null assertion (!)
+		) {
 			return true;
 		}
-		current = current.parent;
-	}
-	return false;
-}
-
-// Helper function to get the type of a node
-function getNodeType(node, parserServices) {
-	const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
-	const type = parserServices.program.getTypeChecker().getTypeAtLocation(tsNode);
-	return type;
-}
-
-// Helper function to determine if a node is defined. This has all the cases which define
-function isDefined(node) {
-	if (!node.parent) return false;
-	if (node.optional === true || node.parent.type === "TSNonNullExpression") return true;
 
-	let current = node;
-	while (current && current.parent) {
-		// Handle if statements that check property existence
-		if (current.parent.type === "IfStatement" && current.parent.test) {
-			const base = getBaseObject(node);
-			if (base && current.parent.test.type === "MemberExpression") {
-				const testBase = getBaseObject(current.parent.test);
-				if (base === testBase) {
-					return true;
-				}
-			}
-		}
+		const parent = current.parent;
+		if (parent === null) break;
 
-		// Handle for...of with Object.entries/keys
-		if (current.parent.type === "ForOfStatement") {
-			const right = current.parent.right;
+		// Handle Object.entries/keys loops
+		if (parent.type === "ForOfStatement") {
+			const right = parent.right;
 			if (
-				right.type === "CallExpression" &&
-				right.callee.type === "MemberExpression" &&
+				right?.type === "CallExpression" &&
+				right.callee?.type === "MemberExpression" &&
 				right.callee.object.name === "Object" &&
 				(right.callee.property.name === "entries" || right.callee.property.name === "keys")
 			) {
-				const base = getBaseObject(node);
-				if (right.arguments[0].type === "Identifier" && right.arguments[0].name === base) {
+				return true;
+			}
+		}
+
+		// Handle presence checks
+		if (parent.type === "IfStatement" || parent.type === "ConditionalExpression") {
+			if (parent.test) {
```

Comment:
Is there a scenario where `parent.test` returns false for an if or a conditional expression? Maybe when the developer hasn't finished writing it? If so let's add a note about it. I think I'd also merge it into the parent if to reduce one level of nesting here.